pkg/network/server.go

```diff
@@ -710,9 +714,33 @@ func (s *Server) handleGetAddrCmd(p Peer) error {
 
 // requestBlocks sends a CMDGetBlockByIndex message to the peer
 // to sync up in blocks. A maximum of maxBlockBatch will
-// send at once.
+// send at once. Two things we need to take care of:
+// 1. If possible, blocks should be fetched in parallel.
+//    height..+500 to one peer, height+500..+1000 to another etc.
+// 2. Every block must eventually be fetched even if peer sends no answer.
+// Thus the following algorithm is used:
+// 1. Block range is divided into chunks of payload.MaxHashesCount.
+// 2. Send requests for chunk in increasing order.
+// 3. After all requests were sent, request random height.
 func (s *Server) requestBlocks(p Peer) error {
-	payload := payload.NewGetBlockByIndex(s.chain.BlockHeight()+1, -1)
+	var currHeight = s.chain.BlockHeight()
+	var peerHeight = p.LastBlockIndex()
+	var needHeight uint32
+	var old = s.lastRequestedHeight.Load()
+	if old <= currHeight {
+		needHeight = currHeight + 1
+		s.lastRequestedHeight.CAS(old, needHeight)
```

Comment:
CAS returns a value and you need to handle it to ensure correctness.

Reply:
Fixed.